### DISCUSSION TITLE
How to show help/usage message on default run of cliffy CLI

### DISCUSSION THREAD

**Discussion by grempe** [c4spar/cliffy#179] [Q&A] [answered]
I'm creating a new CLI which I'd like to have display the top level help content when the cli is run with no commands or flags. Right now I'm just `console.log()` a statement that suggests they try to re-run with the `--help` command. This is not a very satisfying user experience though.

The same applies for sub-commands that the user runs with no commands or flags.

As I understand it I'd need to implement an `.action()` on the top level command which would call a function to show the same help as I see with `--help` but I can't seem to find what function I should call to display the help.

Here is an example of what I'm trying to achieve:

```typescript
import { Command } from "https://deno.land/x/cliffy@v0.18.2/command/mod.ts";
import { HelpCommand } from "https://deno.land/x/cliffy@v0.18.2/command/help/mod.ts";

// Question : how to get both of these commands to print the help?
// deno run helpme.ts
// deno run helpme.ts --help
//
// And these also:
// deno run helpme.ts subcommand
// deno run helpme.ts subcommand --help

const subcommand = new Command()
  .description("HelpMe Demo Sub-Command")
  .action(() => {
    console.log(`How can I display the same output as 'deno run helpme.ts subcommand --help' here?`)
    Deno.exit(0);
  });

const cmd = new Command()
  .name("helpme")
  .version("0.0.0")
  .description("HelpMe Demo")
  .action(() => {
    console.log(`How can I display the same output as 'deno run helpme.ts --help' here?`)
    Deno.exit(0);
  })
  .command("subcommand", subcommand)
  .command("help", new HelpCommand().global());

try {
  cmd.parse(Deno.args);
} catch (error) {
  console.error("Error: ", error.message);
  Deno.exit(1);
}
```

Thanks!

**Comment by c4spar** [comment, marked as answer]
Hi @grempe,

you can use `cmd.showHelp()`. It prints the help text to stdout and there is also a `.getHelp()` method which returns the help text as string.

**Reply by grempe** [reply]
Thanks for the quick response. That is exactly what I was looking for.

`subcommand.showHelp()` also worked as expected to show default help for the subcommand with no args.

I did just go through the docs again to see if I could find this documented. I only found:

```txt
You can also override the help output with the .help() method. This overrides the output of the .getHelp() and .showHelp() method's which is used by the -h and --help option and the help command. The help handler will be used for each command, but can be overridden by child commands.
```

This text didn't make it clear that this is another way to use `.showHelp()`. Perhaps this case could be documented, or an example created as I think this is a fairly common way to display usage in CLI's.

Thanks!

**Reply by c4spar** [reply]
Yeah that's true, the documentation is lacking some information. Will update the documentation soon. I'm also planning to build a new website with a better structured documentation.

**Reply by callumgare** [reply]
The docs are still not very clear about this so as a heads up for future readers here's a code snippit:
``` typescript
import { Command } from "https://deno.land/x/cliffy@v1.0.0-rc.4/command/mod.ts";
import subCommand from "./commands/sub-command.ts";
import denoConfig from "./deno.json" with { type: "json" };

export default new Command()
  .name("name-of-command")
  .version(denoConfig.version)
  .description("Description of command")
  // Note: this .action() call seems to need to be before any .command() calls otherwise it's 
  // silently not run (at least that was the case for me).
  .action(function () {
    this.showHelp();
  })
  .command("sub-command", subCommand)
  .parse(Deno.args);
```

**Reply by c4spar** [reply]
> Note: this .action() call seems to need to be before any .command() calls otherwise it's silently not run (at least that was the case for me).

@callumgare if you call the `.action()` method after `.command("sub-command", subCommand)`, then the action is registered to the `sub-command` command and not to the main command.

**Reply by callumgare** [reply]
Ahh, good to know. Ta :)

**Comment by hojongs** [comment]
I had the same issue and this discussion solved my issue! Very helpful!
